docs/architecture/adr-104-out-of-band-state-sync.md

```diff
@@ -0,0 +1,215 @@
+of its weakness.
+In fact, while downloading recent snapshots is very convenient for new nodes
+(clients of the protocol), providing snapshots to multiple peers (as servers of
+the protocol) is _bandwidth-consuming_.
+As a result, the number of nodes in production CometBFT networks offering the
+state sync service (i.e., servers offering snapshots) has been limited, which
+has rendered the service _fragile_ (from the client's point of view).
+
+> 3. High level idea behind the solution
+
+This ADR stems from the observation that state sync is more than a protocol to
+discover and fetch snapshots from peers in the network.
+In fact, in addition to installing a snapshot, state sync also checks the
+consistency of the installed application state with the state (`appHash`)
+recorded on the blockchain at the same height, and bootstraps CometBFT's block
+store and state store accordingly.
+As a result, once state sync is completed the node can safely switch to Block
+sync and/or Consensus to catch up with the latest state of the network.
+
+The purpose of this ADR is to provide node operators with more flexibility in
+defining how or where state sync should look for application snapshots.
+More precisely, it enables state sync to support the bootstrap of nodes from
+application snapshots obtained _out-of-band_ by operators.
+The goal is to provide an alternative to the mechanism currently adopted by
+state sync, discovering and fetching application snapshots from peers in the
+network, in order to address the above mentioned limitations, while preserving
+most of state sync's operation.
+
+## Alternative Approaches
+
+> This section contains information around alternative options that are considered
+> before making a decision. It should contain a explanation on why the alternative
+> approach(es) were not chosen.
+
+### Strengthen p2p-based state sync
+
+This ADR proposes a workaround for some limitations of the existing state sync
+protocol, briefly summarized above. A more comprehensive and sustainable
+long-term solution would require addressing such limitations.
+
+From a protocol design point of view, probably the major limitation of state
+sync is the lack of mechanisms to incentivize and reward established nodes that
+provide "good" (updated and consistent) snapshots to peers.
+The existing protocol is essentially altruistic: it assumes that established
+nodes will support the bootstrap of fresh nodes without receiving anything in
+return, other than having new peers joining their network.
+
+From an operational point of view, the design of the p2p layer should take into
+consideration the fact that not every node in the network is a "good" state
+sync server. Fresh nodes, which need support for bootstrapping, should then be
+able to discover peers in the network that are able or willing to provide
+application snapshots.
+The implementation of this feature would require a more complex content-based
```

Comment:
While I agree in principle with this statement, I am wondering, do we know whether in practice, at this moment, the set of nodes contacted via RPC for statesync is overlapping with the peers the node connects in any case? 

My reasoning is as follows:
1. If it overlaps, then we contact those peers anyways, and we could just have a peer signal whether it is also a statesync peer.. and then a node keeps track of those peers. 
2. If it does not, or it overlaps partially,we still need to pre-configure the set of peers to use for statesync, or at least have a special statesync seed node maybe that would give us useful peers for statesync. This might be a better/easier solution or complimentary to 1. because we still need to have enough statesyncing peers to be able to light client verify the headers.

Reply:
This is a good point, but I have not idea. In principle, nodes that we connect via RPC to implement the light client should have no relation with the peers a node has in the p2p layer. I didn't fully follow you reasoning from this point, though.

Reply:
Thanks! I know that in principle they do not have to be related, was just wondering whether in practice they are. 

My other point was more with regards to the discussion on strengthening the p2p based statesync. We would still need to have a user manually configure state syncing peers or have a new special node that serves as a seed for statesyncing peers. The node discovers peers who provide snapshots and gossips those around.